streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -71,6 +71,48 @@
     private static final Logger log = LoggerFactory.getLogger(StreamThread.class);
     private static final AtomicInteger STREAM_THREAD_ID_SEQUENCE = new AtomicInteger(1);
 
+    /**
+     * Stream thread states are the possible state that a stream thread can be in.
+     * A thread should only be in one state at a time
+     * The expected state transition with the following defined states is:
+     *
+     *                +-----------+
+     *                |Not Running|
+     *                +-----+-----+
+     *                      |
+     *                      v
+     *                +-----+-----+
+     *          +---->| Running   |<------------+
+     *          |     +-----+-----+             |
+     *          |           |                   |
+     *          |           v                   |
+     *          |     +-----+------------+      |
+     *          <---- |Partitions        |      |
+     *          |     |Revoked           |      |
+     *          |     +-----+------------+      |
+     *          |           |                   |
+     *          |           v                   |
+     *          |     +-----+------------+      |
+     *          |     |Partitions        |      |
+     *          |     |Assigned          |------+
+     *          |     +-----+------------+
+     *          |
+     *          |
+     *          |    +-----+----------+
+     *          +--->|Pending         |
+     *               |Shutdown        |
+     *               +-----+----------+
+     *                     |
+     *                     v
+     *               +-----+-----+
+     *               |Not Running|
+     *               +-----------+
+     *
+     * Custom states is also allowed for cases where there are custom kafka states for different scenarios.
+     */
+    public enum StreamThreadStateType { NOT_RUNNING, RUNNING, PARTITIONS_REVOKED, PARTITIONS_ASSIGNED, PENDING_SHUTDOWN }
```

Comment:
similar to @ijuma's suggestion above we might also want to call this `StreamThread.State` to be consistent with `KafkaStreams.State`